### PR TITLE
[FEATURE] Introduce SecretStr and SecretBytes classes for enhanced secret handling

### LINKED ISSUE #165: [FEATURE] Bigger, better, SecretStr
We should create a custom SecretStr implementation that plays a bit nicer with str operations, but that's up for debate.

For example, this code would throw an error:

```python
from pydantic import SecretStr
body = SecretStr("foobarbazbladibla")
type = "bearer"
header = type + body
```
results in `TypeError: unsupported operand type(s) for +: 'SecretStr' and 'str'`

using f-string notation also doesn't work nicely:

```python
header = f"{type} {body}"
```
you don't get an error doing this, but the resulting type of header will be str and will be set to `'bearer **********'`. The resulting str also seems to be 'broken' in that you can't run type against it for example.

Long story short, we should make 'adding' things (prefixing of suffixes) something to a given SecretStr easy as well.

### PR DESCRIPTION
**Pull request by dannymeijer**
Introduces an extended `SecretStr` class to allow concatenation and use in f-strings along with a similarily enhanced `SecretBytes` class.

## Description
This pull request introduces the `SecretStr` and `SecretBytes` classes, which extends Pydantic's `SecretStr` and `SecretBytes` to provide additional functionality. 

The enhancements include:
- Support for concatenation with other strings and `SecretStr` instances.
- Support for concatenation with same types and `SecretBytes` instances.
- Advanced f-string formatting support for `SecretStr` to ensure the secret value is only revealed in secure contexts.

The `SecretStr` and `SecretBytes` classes continue to ensure that the secret value is masked in string representations and provides methods to safely handle secret values.

## Related Issue
Closes #165

## Motivation and Context
The change is required to provide a more secure and flexible way to handle secret strings while allowing them to be combined with other strings and f-string. By extending Pydantic's `SecretStr`, we can ensure that secret values are properly masked and only revealed in secure contexts, such as within another `SecretStr` instance. This helps prevent accidental exposure of sensitive information. A similar extension was made for Pydantic's `SecretBytes` as well.

A `SecretMixin` class was created to reduce code duplication across `SecretByte` and `SecretStr`.

## How Has This Been Tested?
The changes have been tested by adding unit tests to verify the following:
- The integrity of the `str` and `repr` methods in `SecretStr` and `SecretByte`.
- Concatenation of `SecretStr` with various types, including strings, integers, floats, booleans, and more.
- Advanced f-string formatting to ensure the secret value is only revealed in secure contexts.
- Handling of edge cases where concatenation or string conversion might fail.

## Screenshots (if appropriate):
N/A

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/koheesio/models/__init__.py [hunk 1]
```diff
@@ -762,3 +763,211 @@ def _list_of_columns_validation(columns_value: Union[str, list]) -> list:
...
+            # - stack[1][4][0]: the string representation of the line of code that called this function
+        except IndexError:
+            caller_context = ""
+
+        # safe context: the secret value is returned
+        if 'SecretStr(f' in caller_context:
+            return self.get_secret_value()
+
+        # unsafe context: we let pydantic handle the formatting
+        return super().__format__(format_spec)
```
Comment (riccamini) [inline src/koheesio/models/__init__.py:896]: If I get this right: if we are calling __format__ from within the class, this means that we are safe to use the secret - not that will be showed in any way, but it can serve as for example in concatenation operation. If not, will rely on the super method, which will get the string representation calling str(secret), and ending up with ******.

One curiosity: was this needed indeed for concatenation and the other operations that you added support for or it, or it is just to have a base that can be extended for other operations too?

**Review by riccamini — APPROVED** [review]
Nice addition, I see some use cases for managing parts of the same secret in an independent yet secure way 💪

File: src/koheesio/models/__init__.py (hunk 1 above)
Comment (dannymeijer) [reply src/koheesio/models/__init__.py:896]: the idea is that you can do string concatenation by using f-string, but ONLY if you wrap it in a `SecretStr` call. So like this:

```python
my_super_secret_string = SecretStr("I am very secret")
add_a_prefix = SecretStr(f"A prefix, {my_super_secret_string}")
```

add_a_prefix will now be 'A prefix, I am very secret' that is still a SecretStr. Without this you would have to explicitly call `my_super_secret_string.get_secret_value()`.

Does that make sense?

**Review by pariksheet — APPROVED** [review]
super

**Comment by dannymeijer** [comment]
I had to make some changes to properly accommodate Python 3.12 - please re-review

File: src/koheesio/models/__init__.py [hunk 2]
```diff
@@ -764,6 +767,242 @@ def _list_of_columns_validation(columns_value: Union[str, list]) -> list:
...
+    If one of the list is wrapped in a SecretBytes instance, the '+' operator will work, just like the example above:
+    
+    ```python
+    list1 = SecretBytes([1,2,3])
+    list2 = [4,5,6]
+    list1 + list2  # SecretBytes([1, 2, 3, 4, 5, 6])
+    ```
+
+    If however you try to add any other type of data to the `list1` from the example above, you will get a `TypeError`.
+    You can use the `secret_data_type` method to check the type of the secret data stored in the `SecretBytes` instance.
```
Comment (riccamini) [inline src/koheesio/models/__init__.py:988]: Trying to understand why this is the case 🔬

File: tests/models/test_models.py [hunk 3]
```diff
@@ -253,6 +292,299 @@ class SimpleModelWithExtraParams(BaseModel, ExtraParamsMixin):
...
+        non_secure_context = f"!@#$%^&*({input_str})"
+        assert non_secure_context == "!@#$%^&*(**********)"
+
+        # f-string with escaped characters - secure context
+        escaped_chars_f_string = SecretStr(f"\\n{input_str}\\t")
+        assert escaped_chars_f_string.get_secret_value() == "\\nmy_super_secret\\t"
+
+        # f-string with escaped characters - non secure context
+        non_secure_context = f"\\n{input_str}\\t"
+        assert non_secure_context == "\\n**********\\t"
```
Comment (riccamini) [inline tests/models/test_models.py:514]: Dope coverage for edge cases 🔥

**Review by riccamini — APPROVED** [review]
Nice 💪 Just one question for my personal curiosity

File: src/koheesio/models/__init__.py (hunk 2 above)
Comment (dannymeijer) [reply src/koheesio/models/__init__.py:988]: Standard python behavior. 

This for example will throw a `TypeError: can only concatenate list (not "str") to list`:
```python
some_list = ["1", "2", "3"]
some_list + "5"
```

It's the same concept.

**Review by pariksheet — APPROVED** [review]
LGTM
